compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1027,13 +1025,13 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                     //     fn foo(x: dyn Iterator<Item = impl Debug>)
                     ImplTraitContext::Universal if self.is_in_dyn_type => (true, itctx),
 
-                    // In `type Foo = dyn Iterator<Item: Debug>` we desugar to
-                    // `type Foo = dyn Iterator<Item = impl Debug>` but we have to override the
-                    // "impl trait context" to permit `impl Debug` in this position (it desugars
-                    // then to an opaque type).
-                    //
-                    // FIXME: this is only needed until `impl Trait` is allowed in type aliases.
-                    ImplTraitContext::Disallowed(_) if self.is_in_dyn_type => (true, itctx_tait),
+                    ImplTraitContext::Disallowed(position) if self.is_in_dyn_type => {
+                        self.tcx.sess.emit_err(errors::MisplacedAssocTyBinding {
+                            span: constraint.span,
+                            position: DiagnosticArgFromDisplay(position),
+                        });
+                        (false, itctx)
```

Comment:
We can now stop assigning `itctx` in this match, and just yield true/false for `desugar_to_impl_trait`.